### PR TITLE
BUG: read_excel with openpyxl produces trailing rows of nan

### LINKED ISSUE #39181: BUG: ExcelWriter.book --> no member ->Slow Execution Time<-
- [*] I have checked that this issue has not already been reported.

- [*] I have confirmed this bug exists on the latest version of pandas.

- [*] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

Edit: As indicated in https://github.com/pandas-dev/pandas/issues/39181#issuecomment-769507379, the source of the performance degradation was trailing nans at the end of the DataFrame after reading - rhshadrach

#### Code Sample, a copy-pastable example

```python
# Your code here

writer = ExcelWriter(# pylint: disable=abstract-class-instantiated
        path=fileName, engine='xlsxwriter')  
    workbook = writer.book # pylint: disable=no-member
    header_format = workbook.add_format({'bold': True, 'align': 'center',
                                         'fg_color': '#80bfff', 'border': 2,
                                          'font_name': 'Times New Roman', 'font_size': 9})

#### Problem description
the pylint detects that ExcelWriter don't have "book" member, but the code works well just have 
a very bad execution time in comparison with the 1.1.5 Pandas version 

#### Expected Output
Save DataFrames in the same file but in different sheets.
Expected time: 30 sec   (pandas==1.1.5)
Current time: 90 sec    (pandas==1.2.0)

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 3e89b4c4b1580aa890023fc550774e63d499da25
python           : 3.8.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Spanish_Mexico.1252

pandas           : 1.2.0
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 49.2.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : 3.2.1
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.6.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.1
numexpr          : None
odfpy            : None
openpyxl         : 3.0.5
pandas_gbq       : None
pyarrow          : 2.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.5.4
sqlalchemy       : None
tables           : None
tabulate         : 0.8.7
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #39181
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

This is on top of #39486

### CODE REVIEW

**Comment by jreback** [comment]
why is this not an upstream bug? (not averse to fixing in pandas at least tactically).

**Comment by rhshadrach** [comment]
@jreback [openpyxl docs](https://openpyxl.readthedocs.io/en/stable/optimized.html):

> Read-only mode relies on applications and libraries that created the file providing correct information about the worksheets, specifically the used part of it, known as the dimensions. Some applications set this incorrectly. You can check the apparent dimensions of a worksheet using ws.calculate_dimension(). If this returns a range that you know is incorrect, say A1:A1 then simply resetting the max_row and max_column attributes should allow you to work with the file:

We could fix this issue by calling `sheet.calculate_dimension(force=True)` but this involves a 50% performance hit (measured on a frame with 10000 rows and 10 columns). The method used here involves a 2% performance hit.

**Comment by jreback** [comment]
> @jreback [openpyxl docs](https://openpyxl.readthedocs.io/en/stable/optimized.html):
> 
> > Read-only mode relies on applications and libraries that created the file providing correct information about the worksheets, specifically the used part of it, known as the dimensions. Some applications set this incorrectly. You can check the apparent dimensions of a worksheet using ws.calculate_dimension(). If this returns a range that you know is incorrect, say A1:A1 then simply resetting the max_row and max_column attributes should allow you to work with the file:
> 
> We could fix this issue by calling `sheet.calculate_dimension(force=True)` but this involves a 50% performance hit (measured on a frame with 10000 rows and 10 columns). The method used here involves a 2% performance hit. Ref: [#39001 (comment)](https://github.com/pandas-dev/pandas/issues/39001#issuecomment-770277269)

ok! yeah as long as its robust ok with this

File: pandas/io/excel/_openpyxl.py [hunk 1]
```diff
@@ -525,8 +525,28 @@ def _convert_cell(self, cell, convert_float: bool) -> Scalar:
...
+            if not all(cell == "" for cell in converted_row):
+                last_row_with_data = row_number
+            data.append(converted_row)
+
+        if len(data) > 0:
+            # Trim trailing empty rows
+            data = data[: last_row_with_data + 1]
+
+            # With dimension reset, openpyxl no longer pads rows
+            max_width = max(len(data_row) for data_row in data)
```
Comment (Barben360) [inline pandas/io/excel/_openpyxl.py:546]: With some xlsx files I get an error:

`max() arg is an empty sequence`

I am pretty sure this happens when data object is empty

Comment (Barben360) [reply pandas/io/excel/_openpyxl.py:546]: I can't share the file but I can tell it starts with empty rows and empty columns.

Comment (rhshadrach) [reply pandas/io/excel/_openpyxl.py:546]: Yep - thanks for catching this!

**Comment by jreback** [comment]
lgtm. @simonjayhawkins prob worth doing for 1.2.2

**Comment by simonjayhawkins** [comment]
sure

**Comment by jreback** [comment]
```
==================================== ERRORS ====================================
__________ ERROR collecting scripts/tests/test_validate_docstrings.py __________
ImportError while importing test module '/home/runner/work/pandas/pandas/scripts/tests/test_validate_docstrings.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/share/miniconda/envs/pandas-dev/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
scripts/tests/test_validate_docstrings.py:5: in <module>
    import validate_docstrings
scripts/validate_docstrings.py:50: in <module>
    from numpydoc.validate import validate, Docstring  # isort:skip
E   ImportError: cannot import name 'Docstring' from 'numpydoc.validate' (/usr/share/miniconda/envs/pandas-dev/lib/python3.8/site-packages/numpydoc/validate.py)
=========================== short test summary info ============================
ERROR scripts/tests/test_validate_docstrings.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.54s ===============================
```

hmm is this a suprious issue on the ci / checks?

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.2.x

**Pull request merged**